src/core/i18n/translate.js

```diff
@@ -49,6 +44,12 @@ export default function translate(options = {}) {
 
     Translate.displayName = `Translate[${getDisplayName(WrappedComponent)}]`;
 
-    return Translate;
+    function mapStateToProps({ api, i18n }) {
+      return {
+        i18n: makeI18n(i18n, api.lang),
```

Comment:
I don't think we'd need to store the i18n object in state, we only need the `lang` and `i18nData` in state. By connecting to those directly we would only re-render when their values change [1]. We could expose the `i18n` prop in a HOC by decorating every component that needs it with something like this:
````javascript
function withI18nProp(WrappedComponent) {
  function Localizer(props) {
    const { i18nData, lang, ...otherProps } = props;
    return <WrappedComponent i18n={makeI18n(i18nData, lang)} {...otherProps} />;
  }

  function mapStateToProps(state) {
    // Only re-render when lang or i18nData changes:
    return { lang: state.api.lang, i18nData: state.api.i18nData },
  }

  return compose(
    connect(mapStateToProps),
  )(Localizer);
}
````
[1] I don't think redux will descend into an object to diff all its keys so putting a complete `i18n` object in state might cause problems.